### ISSUE TITLE
TimeZone awareness for CronJob

### ISSUE THREAD

**Issue by schmitch** [jamesmh/coravel#143]
**Describe the solution you'd like**
currently it is impossible to schedule CronJob's with a different timezone that has a daylight saving time to always be on the "local" time provided.
It would be great if the Cron function can also have an additional paramenter `TimeZoneInfo`.

Btw. there was also a Issue that wanted to make the Scheduler TimeZone aware: #75, but I think it's better to specify if the TimeZone provided was either UTC or anything else.

**Comment by jamesmh** [comment]
I vowed that I would always make coravel stick with UTC lol

Will have to think about this.... I suppose daylight savings time would be one argument in favor?

**Comment by schmitch** [comment]
@jamesmh I created a PR about that. Maybe you can look into it. The implementation favors your "always UTC", expect where it would break. i.e. use UTC everywhere and only give a "this was created on timezone X"

**Comment by jamesmh** [comment]
Just a saw that 😀 I'll get back to you once I have some time to take a look 👍

**Comment by jamesmh** [comment]
So one thing is the API:

Let say I do `Cron("00 13 * * *")` (run at 1pm daily).

If I did `CronWithTimeZone("00 13 * * *", timezone)` -> should I expect:

a. That time supplied is UTC but will be converted to the timezone and therefore take into account the timezone DST?
b. The time I supply is expected to be the time "inside" or "for" the timezone?

I feel like option b. would be more appropriate / self-evident, vs. passing in a UTC time and having to "know" that coravel converts.

Does that makes sense?

**Comment by schmitch** [comment]
it's option b. if I run `CronWithTimeZone("00 13 * * *", TimeZoneInfo.FindSystemTimeZoneById("Central European Standard Time"))`* 
it will **always** run on 0:13 CET/CEST. So in winters the cronjob will run at 23:13 UTC and in summer it will run at 22:13 UTC. 

I think option a. is also not possible since DST's exist. I mean when I'm at 15:00 CET I want the cronjob to also happen on 15:00 CEST, because that's how we humans think. But since Coravel is always in UTC, I need to provide him with a info (the time zone that I used for scheduling) so that it works the "human" way.

option c. would be the one I would prefer, remove DST from our timezones, but sadly that won't happen that soon.

So basically the API works like b. if I call `CronWithTimeZone("00 00 * * *", TimeZoneInfo.FindSystemTimeZoneById("Central European Standard Time"))`* it will always run at 0:00 CET/CEST, so the time in UTC is different in summer.

* `TimeZoneInfo.FindSystemTimeZoneById` works differently in Windows/*nix: https://devblogs.microsoft.com/dotnet/cross-platform-time-zones-with-net-core/ that's why there is a helpful package to have a cross-plat way: `TimeZoneConverter`

**Comment by jamesmh** [comment]
I closed the PR for now. My thoughts after coming back to this are to avoid TZ altogether. It's a trade-off. Some people might not want to use coravel because of this, and I suppose something like hangfire might be better in that event?

It's a tough one, because I get that DST affects schedules. But one of the first principles I decided on when I started this project was no TZ logic lol. Things change and opinions change, but I think that's one that I need to stick to.

Thanks for contributing and working together on this. It's a tough one 😌

**Comment by schmitch** [comment]
well the problem is not that they can't use coravel for using facing stuff without that. not that they do not wan't to use it.
it also does not have any tz logic at all. it only converts Time A provided to the correct zoned datetime, the pr was really small. it's sad that this will not be supported.

**Comment by jamesmh** [comment]
I implemented this, but in a quite different approach. Now you can zone any of coravel's schedule methods:

https://docs.coravel.net/Scheduler/#zoned-schedules

Hope that helps with what you need!
